Enhancement to server encoder - NvPipe integration

anderm opened the pull request:
- Added NvPipe library - https://github.com/anderm/NvPipe/tree/low-latency-optimization forked from https://github.com/NVIDIA/NvPipe 
- Integrated NvPipe into WebRTC H264 encoder implementation and removed old Nvencode headers
- Added OpenGL sample server 
- Removed software encoder config flag - encoder will now automatically select encoder type based on hardware
- Added Unit Tests for native servers - testing encoder 
- Removed local webrtc CI test - unit tests are now part of the solution

tylergibson commented:
Three issues need resolving before I can fully test:
1) No documentation on nvpipe build process - can't replicate the build
2) All servers crash after disconnecting from an active session - encoder isn't being properly destroyed
3) Can't reconnect any server after a single session (because it crashes)
4) Servers crash on close in debug builds
5) nvToolsExt64_1.dll isn't properly handled.  It should be found via NVTOOLSEXT_PATH or should be hard copied to output directory during build.
6) Two of the 4 NativeServersUnitTests fail on both debug and release x64
7) .runsettings file is missing for unit tests https://msdn.microsoft.com/en-us/library/jj635153.aspx

anderm commented:
Thanks Tyler, this PR is not ready yet. I added a DO NOT MERGE flag (started this PR to keep track of progress) 
1) Agreed - it depends on CMAKE and it's not a very obvious process. I want to add a lot of documentation around Nvpipe, requirements and how to set the proper paths
2, 3 and 4) That is not entirely true - reconnecting is an issue but it's not the encoder - the encoder is destroyed properly, as tested by the unit tests. This issue might be related to the new frame capture architecture - will debug this with Phong.
5) see 1
6) The tests do not fail, it's the environment setup that fails. After you add the proper paths, it will work fine. See 1 and 5. I also need to update the CI build to include the new CUDA paths.
7) The runsettings file is not missing, you can find it here: 3dtoolkit\Samples\Server\NativeServersUnitTests\tests.runsettings

tylergibson commented:
Thanks @anderm - Sounds like we aren't too far away.

Regarding 1, 5, 6 - We should not be relying on setting environment variables, especially for running unit tests on build outputs.  If these libraries aren't put in the path with a default driver/cuda install, we should be copying them local to the executable during build to ensure they get bundled with the executable.  It'll make CI more reliable, packing for deployment easier, and avoids the need for admin privileges to set paths.

7 - Is there anyway to make this the default?  It didn't get picked up automatically and defaulted to running as x86, which of course failed.

If we can fix the crash, copy libs to output on build for cuda, and you can plop in a basic set of  build instructions for nvpipe that I can replicated, LGTM :) 👍

phongcao commented:
Awesome work, @anderm! There are two main issues that we need to fix: 

1. **nvToolsExt64_1.dll** is missing and makes our plugin fail to load. This is what Tyler has mentioned above.
2. By default, installing CUDA Toolkit 9.1 replaces the current drivers by the older ones and this causes weird issues with encoding (unstable image, high latency...). We should mention the correct way of installing necessary drivers and tools in the docs.

anderm commented:
@tylergibson The path issues are now fixed. NvPipe is now loaded dynamically within WebRTC and does not require certain CUDA dlls. Nvencode will work if the required dll's are present, if not, it will switch to software encoding. We are redistributing the cudart64_91.dll and nvToolsExt64_1.dll as part of our setup.cmd and copying them with the executables, we no longer need to install CUDA Toolkit to run our servers. Please give a try and let me know.
7 -  I have not found any way to automatically select the file in Visual Studio. Our CI build works but unfortunately, it's still a manual process inside Visual Studio. 

@phongcao See my comments above, the issues you mention are now fixed.  The only remaining issue is reconnecting, it would be really great if you could have a look as well!

phongcao commented:
@anderm **StreamingDirectxClient** crashes when running in debug mode. Do you encounter that issue?

![crash](https://user-images.githubusercontent.com/9741323/36548546-942d4638-17be-11e8-9d32-c49fafad1d47.JPG)

anderm commented:
@phongcao no, it works fine, have you downloaded the new webrtc lib files? These got updated in my last commit.

phongcao commented:
Yes, I did a fresh clone. @tylergibson Could you please help test in debug as well?

phongcao commented:
@anderm Sorry, I just realized that I forgot to submit the review comment. Can you please re-check and answer my questions?

anderm commented on 3DStreamingToolKit.sln:
```diff
@@ -1,7 +1,7 @@
-# Visual Studio 15
-VisualStudioVersion = 15.0.26430.4
+# Visual Studio 14
+VisualStudioVersion = 14.0.25420.1
```
This was changed by mistake, I will revert the version

anderm commented on Libraries/NvEncoder/NvEncoder.vcxproj:
```diff
@@ -132,35 +97,6 @@
-      <EnableCOMDATFolding>true</EnableCOMDATFolding>
-      <OptimizeReferences>true</OptimizeReferences>
-    </Link>
-  </ItemDefinitionGroup>
```
I documented this in README but we are moving away from x86 support for NVEncode. This is because x86 is gradually being deprecated/removed from CUDA.

phongcao replied:
Thanks, that makes sense!

anderm commented on Libraries/WebRTC/3dtoolkit_upgrades.patch:
```diff
@@ -0,0 +1,2069 @@
+ }
+ 
+-if (rtc_include_tests) {
++if (true) {
```
Great catch, this should not be "if true". I will revert the change.

anderm commented on Libraries/WebRTC/3dtoolkit_upgrades.patch:
```diff
@@ -0,0 +1,2069 @@
+ }
+ 
+-if (rtc_include_tests) {
++if (true) {
```
Same as above

anderm commented on Libraries/WebRTC/3dtoolkit_upgrades.patch:
```diff
@@ -0,0 +1,2069 @@
++#include "webrtc/base/bind.h"
++#include "webrtc/base/asyncinvoker.h"
++
++#define BITSTREAM_BUFFER_SIZE	2 * 1024 * 1024
```
Yes, I will remove it.

anderm commented on Plugins/NativeServerPlugin/src/directx_buffer_capturer.cpp:
```diff
@@ -105,10 +99,17 @@ void DirectXBufferCapturer::SendFrame(ID3D11Texture2D* frame_buffer, int64_t pre
 	{
-		frame.set_staging_frame_buffer(staging_frame_buffer_.Get());
+		D3D11_MAPPED_SUBRESOURCE mapped;
+		if (SUCCEEDED(d3d_context_.Get()->Map(
```
Not sure I see anything wrong here. You are correct, all references to set_staging_frame_buffer have been removed and we are setting rbga buffers using set_frame_buffer.

anderm replied:
The software encoder still uses the I420 buffers, that stays the same as before.

anderm commented on Plugins/NativeServerPlugin/src/opengl_buffer_capturer.cpp:
```diff
@@ -0,0 +1,193 @@
+		glReadPixels(0, 0, buffer_width, buffer_height, PIXEL_FORMAT, GL_UNSIGNED_BYTE, colorBuffer);
+	}
+
+	if (buffer_width > 4096 || buffer_width > 4096)
```
Great catch! Will fix this.

anderm commented on Plugins/NativeServerPlugin/src/opengl_buffer_capturer.cpp:
```diff
@@ -0,0 +1,193 @@
+	rtc::scoped_refptr<webrtc::I420Buffer> buffer;
+	buffer = webrtc::I420Buffer::Create(buffer_width, buffer_height);
+
+	if (use_software_encoder_)
```
Yes, unfortunately, we still need to set an I420 buffer. The OpenH264 encoder inside WebRTC does not use rbga buffers.

anderm commented on Libraries/WebRTC/3dtoolkit_upgrades.patch:
```diff
@@ -0,0 +1,2069 @@
++		if (m_pNvPipeEncoder)
++		{
++			bufferSize = m_encodeConfig.width * m_encodeConfig.height * 4;
++			pFrameBuffer = new uint8_t[bufferSize];
```
I just spotted this error while testing. I will delete the buffer at release.

tylergibson commented on Samples/Server/Opengl-SpinningCube/SpinningCubeServerGL.vcxproj:
```diff
@@ -0,0 +1,125 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
You've got a path naming issue here.  Case sensitive, Opengl puts the proj file in its own folder on non-windows systems.  Need to move this into OpenGL-SpinningCube.

tylergibson commented:
@phongcao @anderm Just pushed some small build config changes that fixed all the issues you were seeing with runtime failures.  Tested and everything looks good functionally to merge this.  Okay to punt the crash fix to test in Phong's PR

anderm commented:
@phongcao @tylergibson Thanks for your reviews! I just pushed a few changes based on the peer reviews. The repo is now valid and can be merged if everything checks out. I'm ok to push the reconnect issues for the next PR, I will create an issue for that.
